neps/nep-0611.md

```diff
@@ -0,0 +1,233 @@
+    }
+}
+```
+
+The semantics for gas key transactions, at the moment of execution (conversion to receipts), are:
+* A gas key transaction is valid iff all of the following are true:
+  * The public key corresponds to a valid gas key;
+  * The gas key has enough balance to cover the total transaction cost (all costs included in config.rs
+    `tx_cost`; this includes amounts included in `Transfer` actions too);
+  * The nonce ID < total number of nonces for the gas key, and the nonce is a valid nonce for that nonce 
+    ID (per the same nonce checks as access key);
+* When converting the gas key transaction to a receipt, the same logic applies as for access key
+  transactions, except
+  * The transaction cost is deducted from the gas key instead of the account.
+  * The new nonce is written for the specific nonce ID under the gas key.
+
+#### Gas Key Pending Transaction Constraints
+Unlike access key transactions, gas key transactions are not limited in parallelism; they are only limited
+by the amount of gas these transactions consume. Specifically, for a gas key $G$, the total cost of 
+the transactions signed with $G$ in the pending transactions must not exceed the balance of $G$
+(according to the state that the pending transactions are calculated based on).
+
+This constraint should be good enough to cover cases of adding, funding, removing, or withdrawing from gas
+keys as well. For adding and funding, we do need the execution to catch up before the newly available
+balance can be used for pending transactions, which is suboptimal but correct. For withdrawing, the
+action of withdrawing itself goes into the cost of the transaction and therefore is accounted for (and
+note that contract execution cannot create receipts that withdraw from gas keys; only gas key
+transactions can). For deletion, balance in gas keys are not refunded, so although subsequent pending
+transactions may end up failing, the gas committed to those transactions are already burnt, eliminating
+the opportunity for the aforementioned attack.
+
+### Pending Transaction Queue
+
+We would now maintain a new data structure that stores the pending transactions, called the Pending
+Transaction Queue. Although it's conceptually a queue, it is stored as a collection indexed by
+(account ID, transaction type), and further indexed by the block hash. Furthermore, the pending
+transaction queue is stored per block per shard, not as a single data structure. The contents of the queue
+is exactly the pending transactions according to the definition above.
+
+The constraints are enforced as described above; we reiterate it here:
+* There cannot be more than 4 access key transactions under the same account in the queue;
+* There cannot be more gas key transactions under the same gas key in the queue whose total transaction
+  cost exceed the gas key's balance.
+
+The constraints are maintained at the time of chunk production: when producing a chunk, we only accept
```

Comment:
To check the constraints hold, it sounds like this means chunk producers need to include a state witness to validators along with the proposed chunk, just like they do today. Does this work against the idea of SPICE separating the consensus and execution? I suppose maybe if this state witness is very small relative to today then SPICE could still be effective.

Reply:
Ah, good point. We are only addressing the attack surface from (unauthenticated) users, not attacks from chunk producers. Chunk producers today (and in SPICE) are high-stake entities because they are also block producers, and mounting this attack as a chunk producer is not even very effective because you can only affect the chunks you produce. It is also already possible today to simply produce a chunk that is empty.